core/src/main/java/com/oath/oak/InternalOakMap.java

```diff
@@ -317,250 +340,240 @@ V put(K key, V value, Function<ByteBuffer, V> transformer) {
+            if (inTheMiddleOfRebalance(c) || finalizeDeletion(c, lookUp)) {
+                continue;
+            }
+            int oldVersion = lookUp == null ? INVALID_VERSION : -Math.abs(lookUp.version);
 
-        int ei;
-        long oldReference = INVALID_VALUE_REFERENCE;
-        if (lookUp != null) {
-            ei = lookUp.entryIndex;
-            assert ei > 0;
-            oldReference = lookUp.valueReference;
-        } else {
-            ei = c.allocateEntryAndKey(key);
-            if (ei == INVALID_ENTRY_INDEX) {
-                rebalance(c);
-                return put(key, value, transformer);
-            }
-            int prevEi = c.linkEntry(ei, key);
-            if (prevEi != ei) {
-                ei = prevEi;
-                oldReference = c.getValueReference(prevEi);
-                // We can use the old reference *only* when it is DELETED_VALUE,
-                // otherwise operation should restart
-                if (oldReference != INVALID_VALUE_REFERENCE) {
-                    return put(key, value, transformer);
+            int ei;
+            if (lookUp != null) {
+                ei = lookUp.entryIndex;
+                assert ei > 0;
+            } else {
+                ei = c.allocateEntryAndKey(key);
+                if (ei == INVALID_ENTRY_INDEX) {
+                    rebalance(c);
+                    continue;
+                }
+                int prevEi = c.linkEntry(ei, key);
+                if (prevEi != ei) {
+                    c.releaseKey(ei);
+                    if (c.getValueReference(prevEi) != INVALID_VALUE_REFERENCE) {
+                        continue;
+                    }
+                    ei = prevEi;
                 }
             }
-        }
 
-        long newValueReference = c.writeValue(value); // write value in place
+            int[] version = new int[1];
+            long newValueReference = c.writeValue(value, version); // write value in place
 
-        Chunk.OpData opData = new Chunk.OpData(Operation.PUT, ei, newValueReference, oldReference, null);
+            Chunk.OpData opData = new Chunk.OpData(ei, INVALID_VALUE_REFERENCE, newValueReference, oldVersion,
+                    version[0]);
 
-        // publish put
-        if (!c.publish()) {
-            c.releaseValue(newValueReference);
-            rebalance(c);
-            return put(key, value, transformer);
-        }
+            if (!c.publish()) {
+                c.releaseValue(newValueReference);
+                rebalance(c);
+                continue;
+            }
 
-        if (!finishAfterPublishing(opData, c)) {
-            return put(key, value, transformer);
+            if (c.linkValue(opData) != TRUE) {
+                c.releaseValue(newValueReference);
+                c.unpublish();
+            } else {
+                c.unpublish();
+                checkRebalance(c);
+                return null;
+            }
         }
-
-        return null;
```

Comment:
Don't we need to return null at the end?

Reply:
I added an if-else clause.
In case the CAS succeed (the result is TRUE), we return ``null``.
Otherwise, the value is freed, the chunk is rebalanced, and the operation is restarted.